### PR TITLE
File upload

### LINKED ISSUE #1064: File upload
Implement file upload system to be used for `media` objects (see #1062)

Use same approach as in BE3 http://bedita.readthedocs.io/en/latest/endpoints/files.html ?

### PR DESCRIPTION
**Pull request by fquffio**
This PR resolves #1064 by introducing file upload to BEdita, which paves the way for a lot of nasty operations. Finally you'll be able to upload that file you downloaded a long time ago using BitTorrent or Emule on BEdita.

### How to upload your favourite movie covered by Copyright on BEdita

The operation is composed of three steps:
 - **Step 1a**: create a Media object
 - **Step 1b**: upload stream
 - **Step 2**: link the stream to the media

While **Step 1a** and **Step 1b** can be performed in parallel, clients **MUST** wait for successful responses of both before attempting to perform **Step 2**.

#### Step 1a: create a Media object
```http
POST /media HTTP/1.1
Authorization: Bearer {{json web token here}}
Accept: application/vnd.api+json
Content-Type: application/vnd.api+json

{
    "data": {
        "type": "media",
        "data": {
            "title": "My favourite movie"
        }
    }
}
```

```http
HTTP/1.1 201 Created
Location: https://api.example.org/media/666
Content-Type: application/vnd.api+json

{
    "data": {
        "id": "666",
        "type": "media",
        "attributes": {
            // ...
            "uname": "media-my-favourite-movie",
            "title": "My favourite movie"
        },
        "meta": { /* ... */ },
        "relationships": {
            "streams": {
                "links": {
                    "self": "https://api.example.org/media/666/relationships/streams",
                    "related": "https://api.example.org/media/666/streams"
                }
            }
        }
    },
    "links": {
        "home": "https://api.example.org/home",
        "self": "https://api.example.org/media"
    }
}
```

#### Step 1b: upload stream
```http
POST /streams/upload/my-movie.mp4 HTTP/1.1
Authorization: Bearer {{json web token here}}
Accept: application/vnd.api+json
Content-Type: video/mp4

{{binary MPEG4 here}}
```

```http
HTTP/1.1 201 Created
Location: https://api.example.org/streams/6800d046-a9c7-43df-830c-f1e0b930cc38
Content-Type: application/vnd.api+json

{
    "data": {
        "id": "6800d046-a9c7-43df-830c-f1e0b930cc38",
        "type": "streams",
        "attributes": {
            "file_name": "my-movie.mp4",
            "mime_type": "video/mp4"
        },
        "meta": {
            // ...
            "version": 1,
            "file_size": 5368709120,
            "hash_md5": "79bbf9f947de46df5c3a1228ccce86d9",
            "hash_sha1": "2c72dd06f179819c38160c867c2408781f2b6e59",
            "url": null
        },
        "relationships": {
            "object": {
                "links": {
                    "self": "https://api.example.org/streams/6800d046-a9c7-43df-830c-f1e0b930cc38/relationships/object",
                    "related": "https://api.example.org/streams/6800d046-a9c7-43df-830c-f1e0b930cc38/object"
                }
            }
        }
    },
    "links": {
        "home": "https://api.example.org/home",
        "self": "https://api.example.org/streams/upload/my-movie.mp4"
    }
}
```

Please note that at this point Stream's `url` is still `null`. Since the Stream hasn't been associated to an object yet, it is not publicly accessible.

#### Step 2: link the stream to the media
```http
PATCH /streams/6800d046-a9c7-43df-830c-f1e0b930cc38/relationships/object HTTP/1.1
Authorization: Bearer {{json web token here}}
Accept: application/vnd.api+json
Content-Type: application/vnd.api+json

{
    "data": {
        "id": "666",
        "type": "media"
    }
}
```

### Important notes

Streams are considered resources just like Roles, Applications, and any other non-object resources. However, there are some limitations:

 - Streams can't (yet) be updated. Any attempt to issue a `PATCH` request on an existing stream will result in a **403 Forbidden**.
 - Relationship between Streams and Media can be read from both ends (`GET /media/:id/streams` and `GET /streams/:uuid/object`). However, it can be managed _only_ from one: `PATCH /streams/:uuid/object` works as a [to-one relationship](http://jsonapi.org/format/#crud-updating-to-one-relationships), while `POST|PATCH|DELETE /media/:id/streams` will result in **403 Forbidden**.
 - Link between Streams and Media is exposed as a full-featured relationship in order to make Streams and Media somehow independent while sticking to JSON API. This has some drawbacks — for instance, streams can't be managed with the easiness of `date_ranges`. In order to make things slightly easier for clients, Streams are automatically included when fetching Media objects. Therefore, there's no need for `?include=streams` query parameter.

### CODE REVIEW

**Comment by fquffio** [comment]
### FilesystemRegistry

A registry to manage multiple "mountpoints" has been introduced. It relies upon [`league/flysystem`](https://flysystem.thephpleague.com/).

#### Configuring
```php
use BEdita\Core\Filesystem\FilesystemRegistry;
use Cake\Core\Plugin;

FilesystemRegistry::setConfig('default', [
    'className' => 'BEdita/Core.Local',
    'path' => '/my/local/filesystem/path',
    'baseUrl' => 'https://static.example.org/files',
]);

if (Plugin::loaded('BEdita/AWS')) {
    FilesystemRegistry::setConfig('s3', [
        'className' => 'BEdita/AWS.S3',
        'region' => 'eu-west-1',
        'host' => 'my-bucket-name',
        'path' => 'object-prefix/',
        'baseUrl' => 'https://cdn.example.org/object-prefix/',
    ]);
}
```

Examples below assume this configuration.

#### Interacting with files
```php
use BEdita\Core\Filesystem\FilesystemRegistry;

$mountManager = FilesystemRegistry::getMountManager();

$fileExists = $mountManager->has('s3://my-file.png');
if ($fileExists) {
    $mountManager->delete('s3://my-file.png');
}

$stream = fopen('php://memory', 'wb+');
fwrite($stream, 'Gustavo Supporto');
rewind($stream);
$mountManager->putStream('default://myDirectory/head_of_support.txt', $stream);
if (is_resource($stream)) {
    fclose($stream);
}

$stream = $mountManager->readStream('default://myDirectory/head_of_support.txt');
echo stream_get_contents($stream);  // Outputs: Gustavo Supporto
if (is_resource($stream)) {
    fclose($stream);
}
```

See [Flysystem Mount Manager](https://flysystem.thephpleague.com/mount-manager/) for full reference.

#### Obtaining public URL for a file
```php
use BEdita\Core\Filesystem\FilesystemRegistry;

echo FilesystemRegistry::getPublicUrl('default://myDirectory/head_of_support.txt');
// Outputs: https://static.example.org/files/myDirectory/head_of_support.txt
```

File: plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace BEdita\Core\Filesystem;
+
+use Cake\Core\InstanceConfigTrait;
+use League\Flysystem\AdapterInterface;
+
+/**
+ * Filesystem adapter.
+ *
+ * This abstract class is a wrapper around Flysystem adapters, to ensure they can be instantiated
```
Comment (didoda) [inline plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php:22]: Flysystem? :-) Filesystem, I guess... or am I wrong? "Flysystem" is at line 67 and 89 too (`\League\Flysystem\AdapterInterface`) and in `plugins/BEdita/Core/src/Filesystem/FilesystemRegistry.php` lines 21, 22, 23...

File: plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        $this->setConfig($config);
+
+        return true;
+    }
+
+    /**
+     * Get the inner adapter class.
+     *
+     * @return \League\Flysystem\AdapterInterface
```
Comment (didoda) [inline plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php]: Flysystem? view previous comment...

File: plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        return $this->adapter = $adapter;
+    }
+
+    /**
+     * Build the inner adapter from configuration.
+     *
+     * @param array $config Adapter configuration.
+     * @return \League\Flysystem\AdapterInterface
```
Comment (didoda) [inline plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php]: Flysystem... as commented above

File: plugins/BEdita/Core/src/Filesystem/FilesystemRegistry.php [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+ */
+
+namespace BEdita\Core\Filesystem;
+
+use BEdita\Core\Filesystem\Adapter\LocalAdapter;
+use BEdita\Core\SingletonTrait;
+use Cake\Core\App;
+use Cake\Core\ObjectRegistry;
+use Cake\Core\StaticConfigTrait;
+use League\Flysystem\Filesystem;
```
Comment (didoda) [inline plugins/BEdita/Core/src/Filesystem/FilesystemRegistry.php]: Just a doubt about `Flysystem`, as commented before... not a big deal, however

File: plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php (hunk 1 above)
Comment (didoda) [reply plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php:22]: mhm... nevermind. `Flysystem` is fine

Comment (fquffio) [reply plugins/BEdita/Core/src/Filesystem/FilesystemAdapter.php:22]: Yup, Flysystem is actually the name of the abstraction library we use under the hood. Can be misleading at times. 🙃

**Comment by stefanorosanelli** [comment]
Good job! 

Some doubts on `media` / `streams` table:
 * in case of remote media (like classic youtube video) `streams` is supposed to be empty? is it used only directly handled stream files? 
 * if not some attributes like `width` / `height` could be moved to `streams` and we could probably get rid of `media` tables
 * are we going to implement multiple media object types like `Image`, `Video`, `Audio` as extension of `Media` ? should we avoid generic `media` object type like we (should) do with `objects`?

File: plugins/BEdita/API/src/Controller/ResourcesController.php [hunk 5]
```diff
@@ -279,13 +283,18 @@ public function related()
...
-
         $id = $this->request->getParam('id');
         $relationship = $this->request->getParam('relationship');
 
         $association = $this->findAssociation($relationship);
 
+        $allowedMethods = ['get', 'post', 'patch', 'delete'];
+        if ($association instanceof BelongsTo || $association instanceof HasOne) {
+            // For to-one relationship, POST and DELETE are not implemented.
+            $allowedMethods = ['get', 'patch'];
```
Comment (stefanorosanelli) [inline plugins/BEdita/API/src/Controller/ResourcesController.php:294]: this line seems not covered by test (we can do it later)

File: plugins/BEdita/Core/src/Model/Behavior/UploadableBehavior.php [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        if ($contents instanceof StreamInterface) {
+            $contents = $contents->detach();
+        }
+
+        if (is_resource($contents)) {
+            return $manager->putStream($path, $contents);
+        }
+
+        return $manager->put($path, $contents);
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Behavior/UploadableBehavior.php:61]: line not covered by test (but we can do it later)

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+     * In the meanwhile, file size, MD5 and SHA1 hashes are
+     *
+     * @param resource $source Original resource.
+     * @return \Zend\Diactoros\Stream
+     * @throws \InvalidArgumentException Throws an exception if the parameter is not a resource.
+     */
+    protected function createStream($source)
+    {
+        if (!is_resource($source)) {
+            throw new \InvalidArgumentException('Cannot create a stream out of anything but a resource');
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php:169]: line not covered by test (we can do it later)

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+     *
+     * @param string $filesystem Filesystem for which path must be generated.
+     * @param int $subLevels Number of sub-levels to organize files.
+     * @return string
+     */
+    public function filesystemPath($filesystem = 'default', $subLevels = 0)
+    {
+        if (!$this->has('uuid')) {
+            // Generate random UUID. This is needed for path, otherwise we'd let Cake take care of it.
+            $this->uuid = Text::uuid();
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php:91]: line not covered by test

**Review by stefanorosanelli — APPROVED** [review]

**Comment by fquffio** [comment]
@stefanorosanelli Yes, according to this implementation (which is still open to discussion, of course) a "remote" media (e.g. a YouTube video) would exist as a Media object that has no Streams. Streams, on the other hand, are supposed to be a mapping of the physical (...) filesystem, and as such we could have a shell that synchronizes the two by deleting Streams whose referenced object has gone lost.

I'd keep the two concepts distinct, and probably I'd keep the `media` table, since it may come in handy somewhen — or maybe not?

For what concerns code coverage, I reckon it's a bit low. 😞  I'll work to cover some more cases, but some other are quite hard to test since there are no "real-world" scenarios that might actually fall under those conditions... yet — not really [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)-compliant to be honest.

PS: I'm not quite sure whether `width`, `height` and `duration` should stay in `media` or rather be moved to `streams`, though. They surely make sense in `streams`, but such metadata can be useful for remote media as well. 😕

**Comment by stefanorosanelli** [comment]
Yes `width`, `height` and `duration` could belong to both tables.

Probably they're more useful in `streams`. For a remote video or image we could probably use some `custom` or `extra` attributes.

**Comment by fquffio** [comment]
Agreed. I'll move them there in another database migration.

**Review by batopa — APPROVED** [review]

**Pull request merged**